chalice/package.py

```diff
@@ -1,217 +1,285 @@
+                 resource_builder,  # type: ResourceBuilder
+                 post_processor,    # type: TemplatePostProcessor
+                 osutils,           # type: OSUtils
                  ):
         # type: (...) -> None
         self._sam_templater = sam_templater
-        self._lambda_packaager = lambda_packager
+        self._resource_builder = resource_builder
+        self._template_post_processor = post_processor
+        self._osutils = osutils
 
     def _to_json(self, doc):
         # type: (Any) -> str
         return serialize_to_json(doc)
 
-    def package_app(self, config, outdir):
-        # type: (Config, str) -> None
+    def package_app(self, config, outdir, chalice_stage_name):
+        # type: (Config, str, str) -> None
         # Deployment package
-        zip_file = os.path.join(outdir, 'deployment.zip')
-        self._lambda_packaager.create_deployment_package(
-            config.project_dir, config.lambda_python_version, zip_file)
+        resources = self._resource_builder.construct_resources(
+            config, chalice_stage_name)
 
         # SAM template
         sam_template = self._sam_templater.generate_sam_template(
-            config, './deployment.zip')
-        if not os.path.isdir(outdir):
-            os.makedirs(outdir)
-        with open(os.path.join(outdir, 'sam.json'), 'w') as f:
-            f.write(self._to_json(sam_template))
+            resources)
+        if not self._osutils.directory_exists(outdir):
+            self._osutils.makedirs(outdir)
+        self._template_post_processor.process(
+            sam_template, config, outdir, chalice_stage_name)
+        self._osutils.set_file_contents(
+            filename=os.path.join(outdir, 'sam.json'),
+            contents=self._to_json(sam_template),
+            binary=False
+        )
+
+
+class TemplatePostProcessor(object):
+    def __init__(self, osutils):
+        # type: (OSUtils) -> None
+        self._osutils = osutils
+
+    def process(self, template, config, outdir, chalice_stage_name):
+        # type: (Dict[str, Any], Config, str, str) -> None
+        self._fixup_deployment_package(template, outdir)
+
+    def _fixup_deployment_package(self, template, outdir):
+        # type: (Dict[str, Any], str) -> None
+        # NOTE: This isn't my ideal way to do this.  I'd like
+        # to move this into the build step where something
+        # copies the DeploymentPackage.filename over to the
+        # outdir.  That would require plumbing through user
+        # provided params such as "outdir" into the build stage
+        # somehow, which isn't currently possible.
+        for resource in template['Resources'].values():
+            if resource['Type'] != 'AWS::Serverless::Function':
+                continue
+            original_location = resource['Properties']['CodeUri']
+            new_location = os.path.join(outdir, 'deployment.zip')
+            self._osutils.copy(original_location, new_location)
```

Comment:
Are we copying over the `deployment.zip` for every lambda/SAM resource? Is it possible to make sure we do this once only if we are?

Reply:
Yeah we are.  I've updated this.